### ISSUE TITLE
Support for Capacitor 5

### ISSUE THREAD

**Issue by borgoat** [getsentry/sentry-capacitor#355]
Capacitor 5 is around the corner (currently in beta), and we're getting ready for it.
Currently, this plugin is incompatible because of breaking changes due to Android Gradle Plugin 8.

```
A problem occurred configuring project ':sentry-capacitor'.
> Could not create an instance of type com.android.build.api.variant.impl.LibraryVariantBuilderImpl.
   > Namespace not specified. Please specify a namespace in the module's build.gradle file like so:

* Try:
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Exception is:
org.gradle.api.ProjectConfigurationException: A problem occurred configuring project ':sentry-capacitor'.
```

Capacitor has a short guide to migrate plugins to v5. [^1] I can give it a shot and open a MR for it.

[^1]: https://capacitorjs.com/docs/next/updating/plugins/5-0

**Comment by RobSchilderr** [comment]
The impact will be big! Thanks for working on this.

**Comment by HerrDerb** [comment]
It would be very apprechiated if this issue could be resolved 👍 🙂

**Comment by thmclellan** [comment]
@borgoat Thanks for sharing this. Did you have a chance to test this on Android?  I'm seeing a crash on app startup if we call  `Sentry.init( { release, dsn, dist... }, SentryAngular.init)`. This is running from Android Studio on a Pixel with Android 13 (running with Capacitor 5.0.4). Also failing Google Play's internal tests. 

<img width="1243" alt="image" src="https://github.com/getsentry/sentry-capacitor/assets/9682900/e6c72a52-07de-4fc9-9f94-be41416620dd">

@RobSchilderr, I asked on discord, sounds like Lucas is checking this week to validate the impact on the changes to support of capacitor 5 on Capacitor 4 in order to keep it compatible with the new and old capacitor. This is probably a breaking change for Cap 4 due to gradle / java version changes. 

Still troubleshooting, will share here if I can get it working, it's a blocker for publishing our app with Capacitor 5 / Android 13 support.

**Comment by thmclellan** [comment]
It seems the above error was a known issue on sentry-java/sentry-android (https://github.com/getsentry/sentry-java/issues/2601) that was fixed in sentry-java 6.17.0. I was testing with a fork of sentry-capacitor 0.11.4 which had sentry-android 6.11.0. Updating the sentry-capacitor to use sentry-android 6.19.1 fixed the issue. 

I did a test build and it's working on Pixel 4XL with Android 13 and is passing Google Play's startup tests. Using an empty Sentry project, I was able to capture exceptions for web/browser-level errors and native crashes on both Android and iOS (iPhone with iOS 16.4.1). Sentry error capture also worked for web only / non-native.  

@borgoat Thanks again for the initial PR. I had trouble with installing the fork directly because I'm on npm, so I did a modified version with a pre-built dist folder. It's probably best to wait for the real thing, but if anyone wants to try it on a Capacitor 5 project, you can try installing this temporary workaround: `npm i github:thmclellan/sentry-capacitor#cap-v5-npm` and then `npx cap sync` to update plugins.

My package.json sentry entries include: 

```   
    "@sentry/angular": "7.50.0",
    "@sentry/capacitor": "github:thmclellan/sentry-capacitor#cap-v5-npm",
    "@sentry/cli": "2.18.1",
    "@sentry/node": "7.50.0",
    "@sentry/tracing": "7.50.0",
```

Hopefully there are no other problems with moving from sentry-android 6.11.0 to 6.19.1. I'm planning to test further with Android 11 and 12 to see if this is actually stable, and then roll out gradually to a handful of users before doing a mass update. If you're not in a rush it would be best to wait for official support for capacitor 5.

**Comment by RobSchilderr** [comment]
> 

so this doesn't work with yarn?

**Comment by RobSchilderr** [comment]
> It seems the above error was a known issue on sentry-java/sentry-android ([getsentry/sentry-java#2601](https://github.com/getsentry/sentry-java/issues/2601)) that was fixed in sentry-java 6.17.0. I was testing with a fork of sentry-capacitor 0.11.4 which had sentry-android 6.11.0. Updating the sentry-capacitor to use sentry-android 6.19.1 fixed the issue.
> 
> I did a test build and it's working on Pixel 4XL with Android 13 and is passing Google Play's startup tests. Using an empty Sentry project, I was able to capture exceptions for web/browser-level errors and native crashes on both Android and iOS (iPhone with iOS 16.4.1). Sentry error capture also worked for web only / non-native.
> 
> @borgoat Thanks again for the initial PR. I had trouble with installing the fork directly because I'm on npm, so I did a modified version with a pre-built dist folder. It's probably best to wait for the real thing, but if anyone wants to try it on a Capacitor 5 project, you can try installing this temporary workaround: `npm i github:thmclellan/sentry-capacitor#cap-v5-npm` and then `npx cap sync` to update plugins.
> 
> My package.json sentry entries include:
> 
> ```
>     "@sentry/angular": "7.50.0",
>     "@sentry/capacitor": "github:thmclellan/sentry-capacitor#cap-v5-npm",
>     "@sentry/cli": "2.18.1",
>     "@sentry/node": "7.50.0",
>     "@sentry/tracing": "7.50.0",
> ```
> 
> Hopefully there are no other problems with moving from sentry-android 6.11.0 to 6.19.1. I'm planning to test further with Android 11 and 12 to see if this is actually stable, and then roll out gradually to a handful of users before doing a mass update. If you're not in a rush it would be best to wait for official support for capacitor 5.

so this doesn't work with yarn?

**Comment by thmclellan** [comment]
@RobSchilderr It *probably* does work on yarn but I haven't tested it. I just had trouble getting the other fork to work... I assumed it was because I'm using npm for my main Ionic/Capacitor project. 

Yarn has a similar "install from github" option described here (https://womanonrails.com/adding-yarn-package-from-github), so you could try `yarn add github:thmclellan/sentry-capacitor#cap-v5-npm` and `npx cap sync` to synchronize Capacitor plugins. You'll also need the @sentry/angular and @sentry/tracking entries like in my example package.json. Hope that helps - good luck!

**Comment by borgoat** [comment]
@RobSchilderr for yarn, I prepared this other branch with the right configuration in the prepack script. I only tested it with yarn v3 though, I can't guarantee it'll work with other versions...

```json
{
  "dependencies": {
    "@sentry/capacitor": "borgoat/sentry-capacitor#head=yarn-2",
    "@sentry/angular": "7.42.0",
    "@sentry/tracing": "7.42.0",
  }
}
```

@thmclellan I saw you had to actually commit the `dist` folder to make it work with npm. I don't know enough about Node package management to say for sure, but possibly you could also solve that by replacing all the occurrences of "yarn" in this package.json with "npm". The change I had introduced in that branch was already merged here #363 - using `prepack` instead of `prepublishOnly` - to always prepare the `dist` folder on install, and not before publishing it to the npm registry only. But I guess it's not surprising that with that configuration, if you're using npm in the dependent repository, and it finds a `prepack` step invoking yarn, there's not much it can do at that point...

**Comment by thmclellan** [comment]
@borgoat Yes, it got a bit ugly, I would only recommend my fork as a temporary fix for direct installing. I tried implementing a dist build (at time of install) through an npm `postinstall` script, but reading up on `prepack`, I see it is also supported by npm and would have been a cleaner choice. 

When trying the `postinstall` route, I found the package.json `files` section caused several build files to get excluded, and the `.npmignore` was setup to have the same effect. Maybe `prepack` would have solved for that. I captured a related issue https://github.com/getsentry/sentry-capacitor/issues/380.
